src/linux/mod.rs

```diff
@@ -41,14 +42,96 @@ pub struct Segment<'a> {
 }
 
 impl<'a> Segment<'a> {
+    fn phdr(&self) -> &'a Phdr {
+        unsafe { self.phdr.as_ref().unwrap() }
+    }
+
+    /// You must pass this segment's `SharedLibrary` or else this is wild UB.
+    unsafe fn data(&self, shlib: &SharedLibrary<'a>) -> Option<&'a [u8]> {
+        let phdr = self.phdr();
+        (shlib.addr as usize)
+            .checked_add(phdr.p_vaddr as usize)
+            .map(|avma| slice::from_raw_parts(avma as *const u8, phdr.p_memsz as usize))
+    }
+
     fn is_load(&self) -> bool {
-        unsafe {
-            let hdr = self.phdr.as_ref().unwrap();
-            match hdr.p_type {
-                libc::PT_LOAD => true,
-                _ => false,
+        self.phdr().p_type == libc::PT_LOAD
+    }
+
+    fn is_note(&self) -> bool {
+        self.phdr().p_type == libc::PT_NOTE
+    }
+
+    /// Parse the contents of a `PT_NOTE` segment.
+    ///
+    /// Returns a triple of
+    ///
+    /// 1. The `NT_*` note type.
+    /// 2. The note name.
+    /// 3. The note descriptor payload.
+    ///
+    /// You must pass this segment's `SharedLibrary` or else this is wild UB.
+    unsafe fn notes(
+        &self,
+        shlib: &SharedLibrary<'a>,
+    ) -> impl Iterator<Item = (libc::Elf32_Word, &'a [u8], &'a [u8])> {
+        assert!(self.is_note());
+
+        let phdr = self.phdr();
+
+        let alignment = std::cmp::min(phdr.p_align as usize, 4);
+        let align = move |data: &'a [u8]| {
+            // Same logic as in `gimli` which took it from `readelf`.
```

Comment:
Move this comment above the `min`, since that is part of the logic too, and move this `if` outside of the closure, because it only needs to be checked once.
```suggestion
            // Same logic as in `goblin` which took it from `readelf`.
```

Reply:
This closure should get inlined, and then only checked once. The problem with trying to explicitly only check it once is that `alignment` would need to become an `Option<usize>` and then we would be checking the option discriminant, which is equivalent, so it doesn't really buy us anything.